### PR TITLE
feat: support github enterprise api

### PR DESCRIPTION
**Pull request by ricardojdsilva87**
# Pull Request

## Proposed Changes

This pull requests makes possible the authentication against the GitHub Enterprise API with a GitHub Enterprise App using the mentioned environment variables:
- `GH_APP_ID`
- `GH_APP_INSTALLATION_ID`
- `GH_APP_PRIVATE_KEY`

A new environment variable `GITHUB_APP_ENTERPRISE_ONLY` needs to be set to `true` so that the api endpoint can also be adjusted.

If the GHE endpoint is provided a different login is made, also every function checks if the `ghe` variable is set to modify the API endpoint using by default the endpoint `https://api.github.com.
Reusing the existing code a [simple check](https://github.com/github/evergreen/pull/256/files#diff-dbcf43432c07570656f9b214353e5df97b86882c0e6da4bc037be69c25c52174R28-R31) is made to switch the connector on the github library

All the following scenarios should be covered since there was no change in the previous implementation, just a [switch on the login type](https://github.com/github/evergreen/pull/256/files#diff-dbcf43432c07570656f9b214353e5df97b86882c0e6da4bc037be69c25c52174L28-R38) was added for the previous environment variables:
- Token + github.com :white_check_mark: 
- Token + ghe :white_check_mark:
- github app in github.com + github.com :white_check_mark: 
- github app in github.com + ghe :white_check_mark: 
- github app in ghe + ghe :white_check_mark:

Should fix https://github.com/github/issue-metrics/issues/364

This PR also includes some code "cleanup" mostly on the prints to make it easier to read using f"" instead of the + signs.

### Author/Contributor

- [x] If documentation is needed for this change, has that been included in this pull request
- [x] run `make lint` and fix any issues that you have introduced
- [x] run `make test` and ensure you have test coverage for the lines you are introducing
- [ ] If publishing new data to the public (scorecards, security scan results, code quality results, live dashboards, etc.), please request review from `@jeffrey-luszcz`

**make lint results**

```
# stop the build if there are Python syntax errors or undefined names
flake8 . --config=.github/linters/.flake8 --count --select=E9,F63,F7,F82 --show-source
0
# exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
flake8 . --config=.github/linters/.flake8 --count --exit-zero --max-complexity=15 --max-line-length=150
0
isort --settings-file=.github/linters/.isort.cfg .
Skipped 2 files
pylint --rcfile=.github/linters/.python-lint --fail-under=9.0 *.py
************* Module dependabot_file
dependabot_file.py:7:0: R0917: Too many positional arguments (6/5) (too-many-positional-arguments)
dependabot_file.py:52:0: R0917: Too many positional arguments (8/5) (too-many-positional-arguments)
************* Module evergreen
evergreen.py:287:0: R0917: Too many positional arguments (7/5) (too-many-positional-arguments)

------------------------------------------------------------------
Your code has been rated at 9.98/10 (previous run: 9.98/10, +0.00)

mypy --config-file=.github/linters/.mypy.ini *.py
Success: no issues found in 9 source files
black .
reformatted .../evergreen/auth.py
reformatted .../evergreen/test_auth.py
reformatted .../evergreen/evergreen.py
reformatted .../evergreen/test_evergreen.py

All done! ✨ 🍰 ✨
```

**make test results**
```
pytest -v --cov=. --cov-config=.coveragerc --cov-fail-under=80 --cov-report term-missing
================================================================================================================ test session starts ================================================================================================================
platform darwin -- Python 3.13.0, pytest-8.3.3, pluggy-1.5.0 -- /opt/homebrew/opt/python@3.13/bin/python3.13
cachedir: .pytest_cache
rootdir: .../evergreen
plugins: cov-5.0.0
collected 99 items

test_auth.py::TestAuth::test_auth_to_github_with_ghe PASSED                                                                                                                                                                                   [  1%]
test_auth.py::TestAuth::test_auth_to_github_with_token PASSED                                                                                                                                                                                 [  2%]
test_auth.py::TestAuth::test_auth_to_github_without_token PASSED                                                                                                                                                                              [  3%]
test_auth.py::TestAuth::test_get_github_app_installation_token PASSED                                                                                                                                                                         [  4%]
test_dependabot_file.py::TestDependabotFile::test_add_existing_ecosystem_to_exempt_list PASSED                                                                                                                                                [  5%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_for_multiple_repos_with_few_existing_config PASSED                                                                                                                    [  6%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_2_space_indent_existing_config_bundler_with_update PASSED                                                                                                        [  7%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_2_space_indent_existing_config_bundler_with_update_and_no_newline PASSED                                                                                         [  8%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_bundler PASSED                                                                                                                                                   [  9%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_cargo PASSED                                                                                                                                                     [ 10%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_composer PASSED                                                                                                                                                  [ 11%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_docker PASSED                                                                                                                                                    [ 12%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_exempt_ecosystems PASSED                                                                                                                                         [ 13%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_existing_config_bundler_no_update PASSED                                                                                                                         [ 14%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_github_actions PASSED                                                                                                                                            [ 15%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_github_actions_without_files PASSED                                                                                                                              [ 16%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_gomod PASSED                                                                                                                                                     [ 17%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_groups PASSED                                                                                                                                                    [ 18%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_hex PASSED                                                                                                                                                       [ 19%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_label PASSED                                                                                                                                                     [ 20%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_labels PASSED                                                                                                                                                    [ 21%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_maven PASSED                                                                                                                                                     [ 22%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_npm PASSED                                                                                                                                                       [ 23%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_nuget PASSED                                                                                                                                                     [ 24%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_pip PASSED                                                                                                                                                       [ 25%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_repo_specific_exempt_ecosystems PASSED                                                                                                                           [ 26%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_schedule_day PASSED                                                                                                                                              [ 27%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_terraform_with_files PASSED                                                                                                                                      [ 28%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_terraform_without_files PASSED                                                                                                                                   [ 29%]
test_dependabot_file.py::TestDependabotFile::test_build_dependabot_file_with_weird_space_indent_existing_config_bundler_with_update PASSED                                                                                                    [ 30%]
test_dependabot_file.py::TestDependabotFile::test_check_multiple_repos_with_no_dependabot_config PASSED                                                                                                                                       [ 31%]
test_dependabot_file.py::TestDependabotFile::test_not_found_error PASSED                                                                                                                                                                      [ 32%]
test_env.py::TestEnv::test_get_env_vars_auth_with_github_app_installation PASSED                                                                                                                                                              [ 33%]
test_env.py::TestEnv::test_get_env_vars_missing_at_least_one_auth PASSED                                                                                                                                                                      [ 34%]
test_env.py::TestEnv::test_get_env_vars_missing_org_or_repo PASSED                                                                                                                                                                            [ 35%]
test_env.py::TestEnv::test_get_env_vars_optional_values PASSED                                                                                                                                                                                [ 36%]
test_env.py::TestEnv::test_get_env_vars_repo_specific_exemptions_improper_format PASSED                                                                                                                                                       [ 37%]
test_env.py::TestEnv::test_get_env_vars_repo_specific_exemptions_unsupported_ecosystem PASSED                                                                                                                                                 [ 38%]
test_env.py::TestEnv::test_get_env_vars_with_a_valid_label PASSED                                                                                                                                                                             [ 39%]
test_env.py::TestEnv::test_get_env_vars_with_bad_schedule_choice PASSED                                                                                                                                                                       [ 40%]
test_env.py::TestEnv::test_get_env_vars_with_bad_schedule_day_choice PASSED                                                                                                                                                                   [ 41%]
test_env.py::TestEnv::test_get_env_vars_with_badly_formatted_created_after_date PASSED                                                                                                                                                        [ 42%]
test_env.py::TestEnv::test_get_env_vars_with_batch_size PASSED                                                                                                                                                                                [ 43%]
test_env.py::TestEnv::test_get_env_vars_with_invalid_batch_size_int PASSED                                                                                                                                                                    [ 44%]
test_env.py::TestEnv::test_get_env_vars_with_invalid_batch_size_str PASSED                                                                                                                                                                    [ 45%]
test_env.py::TestEnv::test_get_env_vars_with_no_batch_size PASSED                                                                                                                                                                             [ 46%]
test_env.py::TestEnv::test_get_env_vars_with_org PASSED                                                                                                                                                                                       [ 47%]
test_env.py::TestEnv::test_get_env_vars_with_org_and_repo_specific_exemptions PASSED                                                                                                                                                          [ 48%]
test_env.py::TestEnv::test_get_env_vars_with_repos PASSED                                                                                                                                                                                     [ 49%]
test_env.py::TestEnv::test_get_env_vars_with_repos_disabled_security_updates PASSED                                                                                                                                                           [ 50%]
test_env.py::TestEnv::test_get_env_vars_with_repos_exempt_ecosystems PASSED                                                                                                                                                                   [ 51%]
test_env.py::TestEnv::test_get_env_vars_with_repos_filter_visibility_invalid_multiple_value PASSED                                                                                                                                            [ 52%]
test_env.py::TestEnv::test_get_env_vars_with_repos_filter_visibility_invalid_single_value PASSED                                                                                                                                              [ 53%]
test_env.py::TestEnv::test_get_env_vars_with_repos_filter_visibility_multiple_values PASSED                                                                                                                                                   [ 54%]
test_env.py::TestEnv::test_get_env_vars_with_repos_filter_visibility_no_duplicates PASSED                                                                                                                                                     [ 55%]
test_env.py::TestEnv::test_get_env_vars_with_repos_filter_visibility_single_value PASSED                                                                                                                                                      [ 56%]
test_env.py::TestEnv::test_get_env_vars_with_repos_no_dry_run PASSED                                                                                                                                                                          [ 57%]
test_env.py::TestEnv::test_get_env_vars_with_team PASSED                                                                                                                                                                                      [ 58%]
test_env.py::TestEnv::test_get_env_vars_with_team_and_repo PASSED                                                                                                                                                                             [ 59%]
test_env.py::TestEnv::test_get_env_vars_with_update_existing PASSED                                                                                                                                                                           [ 60%]
test_env.py::TestEnv::test_get_env_vars_with_valid_labels_containing_spaces PASSED                                                                                                                                                            [ 61%]
test_env.py::TestEnv::test_get_env_vars_with_valid_schedule_and_schedule_day PASSED                                                                                                                                                           [ 62%]
test_env_get_bool.py::TestEnv::test_get_bool_env_var_that_does_not_exist_and_default_value_returns_false PASSED                                                                                                                               [ 63%]
test_env_get_bool.py::TestEnv::test_get_bool_env_var_that_does_not_exist_and_default_value_returns_true PASSED                                                                                                                                [ 64%]
test_env_get_bool.py::TestEnv::test_get_bool_env_var_that_exists_and_is_false PASSED                                                                                                                                                          [ 65%]
test_env_get_bool.py::TestEnv::test_get_bool_env_var_that_exists_and_is_false_due_to_invalid_value PASSED                                                                                                                                     [ 66%]
test_env_get_bool.py::TestEnv::test_get_bool_env_var_that_exists_and_is_true PASSED                                                                                                                                                           [ 67%]
test_evergreen.py::TestDependabotSecurityUpdates::test_enable_dependabot_security_updates PASSED                                                                                                                                              [ 68%]
test_evergreen.py::TestDependabotSecurityUpdates::test_enable_dependabot_security_updates_failed PASSED                                                                                                                                       [ 69%]
test_evergreen.py::TestDependabotSecurityUpdates::test_is_dependabot_security_updates_disabled PASSED                                                                                                                                         [ 70%]
test_evergreen.py::TestDependabotSecurityUpdates::test_is_dependabot_security_updates_enabled PASSED                                                                                                                                          [ 71%]
test_evergreen.py::TestDependabotSecurityUpdates::test_is_dependabot_security_updates_not_found PASSED                                                                                                                                        [ 72%]
test_evergreen.py::TestCommitChanges::test_commit_changes PASSED                                                                                                                                                                              [ 73%]
test_evergreen.py::TestCheckPendingPullsForDuplicates::test_check_pending_pulls_for_duplicates_no_duplicates PASSED                                                                                                                           [ 74%]
test_evergreen.py::TestCheckPendingPullsForDuplicates::test_check_pending_pulls_for_duplicates_with_duplicates PASSED                                                                                                                         [ 75%]
test_evergreen.py::TestCheckPendingIssuesForDuplicates::test_check_pending_issues_for_duplicates_no_duplicates PASSED                                                                                                                         [ 76%]
test_evergreen.py::TestCheckPendingIssuesForDuplicates::test_check_pending_issues_for_duplicates_with_duplicates PASSED                                                                                                                       [ 77%]
test_evergreen.py::TestGetReposIterator::test_get_repos_iterator_with_organization PASSED                                                                                                                                                     [ 78%]
test_evergreen.py::TestGetReposIterator::test_get_repos_iterator_with_repository_list PASSED                                                                                                                                                  [ 79%]
test_evergreen.py::TestGetReposIterator::test_get_repos_iterator_with_team PASSED                                                                                                                                                             [ 80%]
test_evergreen.py::TestGetGlobalProjectId::test_get_global_project_id_parse_response_failed PASSED                                                                                                                                            [ 81%]
test_evergreen.py::TestGetGlobalProjectId::test_get_global_project_id_request_failed PASSED                                                                                                                                                   [ 82%]
test_evergreen.py::TestGetGlobalProjectId::test_get_global_project_id_success PASSED                                                                                                                                                          [ 83%]
test_evergreen.py::TestGetGlobalIssueId::test_get_global_issue_id_parse_response_failed PASSED                                                                                                                                                [ 84%]
test_evergreen.py::TestGetGlobalIssueId::test_get_global_issue_id_request_failed PASSED                                                                                                                                                       [ 85%]
test_evergreen.py::TestGetGlobalIssueId::test_get_global_issue_id_success PASSED                                                                                                                                                              [ 86%]
test_evergreen.py::TestGetGlobalPullRequestID::test_get_global_pr_id_key_error PASSED                                                                                                                                                         [ 87%]
test_evergreen.py::TestGetGlobalPullRequestID::test_get_global_pr_id_request_exception PASSED                                                                                                                                                 [ 88%]
test_evergreen.py::TestGetGlobalPullRequestID::test_get_global_pr_id_success PASSED                                                                                                                                                           [ 89%]
test_evergreen.py::TestLinkItemToProject::test_link_item_to_project_request_exception PASSED                                                                                                                                                  [ 90%]
test_evergreen.py::TestLinkItemToProject::test_link_item_to_project_success PASSED                                                                                                                                                            [ 91%]
test_evergreen.py::TestIsRepoCreateDateBeforeCreatedAfterDate::test_is_created_after_date_is_empty_string PASSED                                                                                                                              [ 92%]
test_evergreen.py::TestIsRepoCreateDateBeforeCreatedAfterDate::test_is_repo_create_date_before_created_after_date PASSED                                                                                                                      [ 93%]
test_evergreen.py::TestIsRepoCreateDateBeforeCreatedAfterDate::test_is_repo_create_date_is_after_created_after_date PASSED                                                                                                                    [ 94%]
test_evergreen.py::TestIsRepoCreateDateBeforeCreatedAfterDate::test_is_repo_created_date_and_created_after_date_is_not_a_date PASSED                                                                                                          [ 95%]
test_evergreen.py::TestIsRepoCreateDateBeforeCreatedAfterDate::test_is_repo_created_date_has_no_time_zone PASSED                                                                                                                              [ 96%]
test_evergreen.py::TestIsRepoCreateDateBeforeCreatedAfterDate::test_is_repo_created_date_is_before_created_after_date_without_timezene_again PASSED                                                                                           [ 97%]
test_evergreen.py::TestCheckExistingConfig::test_check_existing_config_with_existing_config PASSED                                                                                                                                            [ 98%]
test_evergreen.py::TestCheckExistingConfig::test_check_existing_config_without_existing_config PASSED                                                                                                                                         [100%]

---------- coverage: platform darwin, python 3.13.0-final-0 ----------
Name                 Stmts   Miss  Cover   Missing
--------------------------------------------------
auth.py                 28      9    68%   28-35, 46, 76-78
dependabot_file.py      73      0   100%
env.py                 127     11    91%   42-43, 159-160, 188, 214, 222, 228, 248, 293, 311
evergreen.py           139      3    98%   272-273, 326
--------------------------------------------------
TOTAL                  367     23    94%

Required test coverage of 80% reached. Total coverage: 93.73%

================================================================================================================ 99 passed in 1.47s =================================================================================================================
```

### Reviewer

- [ ] Label as either `fix`, `documentation`, `enhancement`, `infrastructure`, `maintenance` or `breaking`

### CODE REVIEW

**Comment by jmeridth** [comment]
Thank you for your contribution.  Unfortunately we can't assume just because an enterprise url is provided that the GitHub App is also in the enterprise.  It could be outside the enterprise and still need to auth against the default api endpoint.  We need to have the user let us know the app's scope.

As mentioned in [your discussion](https://github.com/github/evergreen/discussions/255#discussioncomment-11029943), allowing apps to be scoped to just the enterprise is a [new feature](https://github.blog/changelog/2024-10-22-enterprises-can-create-github-apps-for-use-within-the-enterprise/).

Thank you for the concatenation to f-strings.  I do agree, that is more readable.

Reviewing your PR now.

**Comment by ricardojdsilva87** [comment]
Hello @jmeridth, thanks for your input,
Regarding that new feature I was not aware of that. We only work with apps created inside of our GHE instance and those can be installed only on that instance organizations.

I've noticed that the github library has 2 different authentication methods depending if it's for github or github enterprise:
image
They might return different values in the end, that might be the reason why the url is needed for ghe login and not for github.com. By default internally it must authenticate against api.github.com.

Like you've mentioned most likely there is the need for that scenario where an app is created outside the GHE and still needs to authenticate against github api, but then the remaining actions must be done against the ghe endpoint. Not sure how to tackle this without more inputs.

Also I have ran prettier against the README file to fix the lint errors. Regarding [these](https://github.com/github/evergreen/actions/runs/11478238635/job/31954048770?pr=256#step:6:414), they were added to put some bullet points in the table, should we add the rule to the ignore list?

Thanks

**Comment by jmeridth** [comment]
> Not sure how to tackle this without more inputs.

Agreed.  I think we'll have to request the user provide another environment variable like `GITHUB_APP_ENTERPRISE_ONLY` (boolean) or something. Naming is hard 😉 Can add to README to ensure the user knows that's available.

> should we add the rule to the ignore list

I'm hesitant to do that but the bulleted list does read better.  Are we able to ignore list the specific elements for `MD033/no-inline-html` or can it only be for the entire `no-inline-html`? 🤔

**Comment by ricardojdsilva87** [comment]
Hello again,
We can ignore the entire rule, superlinter is using [these rules](https://github.com/DavidAnson/markdownlint/blob/main/doc/md033.md), so we can add them in the linter configuration file excluding the elements we need, just [commited that](https://github.com/github/evergreen/pull/256/commits/1408143aa6bf6693837f2efc7d11df51c3ea96ea).
Regarding the new input I will need a bit of time to check what needs to be added, so just to confirm, all the scenarios would be:
- Token + github.com :white_check_mark: 
- Token + ghe :white_check_mark:
- github app in github.com + github.com :question:
- github app in github.com + ghe :question:
- github app in ghe + ghe :white_check_mark:

Those marked with :question: I'm not sure if they are already working in the current code. Also since you know better the code, feel free to commit :)

Thanks

**Comment by jmeridth** [comment]
> Hello again, We can ignore the entire rule, superlinter is using [these rules](https://github.com/DavidAnson/markdownlint/blob/main/doc/md033.md), so we can add them in the linter configuration file excluding the elements we need, just [commited that](https://github.com/github/evergreen/pull/256/commits/1408143aa6bf6693837f2efc7d11df51c3ea96ea). Regarding the new input I will need a bit of time to check what needs to be added, so just to confirm, all the scenarios would be:
> 
> * Token + github.com ✅
> * Token + ghe ✅
> * github app in github.com + github.com ❓
> * github app in github.com + ghe ❓
> * github app in ghe + ghe ✅
> 
> Those marked with ❓ I'm not sure if they are already working in the current code. Also since you know better the code, feel free to do the necessary changes 🙂
> 
> Thanks

Your change handles the last one.  The 3rd and 4th should be covered already but don't mind being double checked.

**Comment by jmeridth** [comment]
Can you please add the new environment variable to `.env-example`?

**Comment by jmeridth** [comment]
> @jmeridth seems that the Docker CI step is stuck for 4 hours. The remaining steps seem to have ran successfully, also no lint issues. Thanks

Yeah, the timeout happens once in a while and a re-run works.  It did again in this case.

Will look at your PR in whole again soon.  When we merge we'll need these additions in the other GitHub OSPO Actions (I can give links), if you're interested.  If not, no worries, I can get them knocked out while pointing here for your kickoff.  Let me know.

**Review by jmeridth — APPROVED** [review]

File: auth.py [hunk 1]
```diff
@@ -1,15 +1,21 @@
 """This is the module that contains functions related to authenticating to GitHub with a personal access token."""
 
+import logging
+import logging.config
+
 import github3
 import requests
 
+logging.basicConfig(level=logging.DEBUG)
+
```
Comment (zkoppert) [inline auth.py]: What information is going to this log? I don't see anywhere that directly references the function i'd expect: [debug()](https://docs.python.org/3/library/logging.html#logging.Logger.debug), [info()](https://docs.python.org/3/library/logging.html#logging.Logger.info), [warning()](https://docs.python.org/3/library/logging.html#logging.Logger.warning), [error()](https://docs.python.org/3/library/logging.html#logging.Logger.error) and [critical()](https://docs.python.org/3/library/logging.html#logging.Logger.critical) methods

Comment (jmeridth) [reply auth.py]: > What information is going to this log? I don't see anywhere that directly references the function i'd expect: [debug()](https://docs.python.org/3/library/logging.html#logging.Logger.debug), [info()](https://docs.python.org/3/library/logging.html#logging.Logger.info), [warning()](https://docs.python.org/3/library/logging.html#logging.Logger.warning), [error()](https://docs.python.org/3/library/logging.html#logging.Logger.error) and [critical()](https://docs.python.org/3/library/logging.html#logging.Logger.critical) methods

🤦 

I believe @ricardojdsilva87 meant to remove this.  They did at one point and it snuck back in.  It was temporary for them.  @ricardojdsilva87 is that correct?  Could you please do a quick follow PR to remove if so?

Comment (jmeridth) [reply auth.py]: [This was the removal commit](9972b96763715d422ab56ee9354966ecdf2c48db)

Comment (jmeridth) [reply auth.py]: Nm.  This was a specific commit.  Final boss version it is not there.  We are good.  My bad.

Comment (zkoppert) [reply auth.py]: Thanks for checking my understanding on this. I'll make sure my review tab is looking at all commits next time.

**Pull request merged**